packages/signer/src/utils.ts

```diff
@@ -0,0 +1,243 @@
+import { sha256 } from "viem";
+import type {
+  AuthProviderConfig,
+  AuthProviderCustomization,
+  GetOauthProviderUrlArgs,
+  KnownAuthProvider,
+  OauthState,
+} from "./types";
+
+const DEFAULT_PROVIDER_CUSTOMIZATION: Record<
+  KnownAuthProvider,
+  AuthProviderCustomization
+> = {
+  google: { scope: "openid email" },
+  apple: { scope: "openid email" },
+  facebook: {
+    scope: "openid email",
+    // Fixes Facebook mobile login so that `window.opener` doesn't get nullified.
+    otherParameters: { sdk: "joey" },
+  },
+  twitch: {
+    scope: "openid user:read:email",
+    claims: JSON.stringify({ id_token: { email: null } }),
+    // Forces Twitch to show the login page even if the user is already logged in.
+    otherParameters: { force_verify: "true" },
+  },
+  auth0: { scope: "openid email" },
+};
+
+/**
+ * Returns the default customization parameters when using a known auth provider
+ *
+ * @param {string} knownAuthProviderId id of a known auth provider, e.g. "google"
+ * @returns {AuthProviderCustomization | undefined} default customization parameters
+ */
+export function getDefaultProviderCustomization(
+  knownAuthProviderId: KnownAuthProvider
+): AuthProviderCustomization | undefined {
+  return DEFAULT_PROVIDER_CUSTOMIZATION[knownAuthProviderId];
+}
+
+/**
+ * Generates an OAuth nonce based on a Turnkey public key
+ *
+ * @param {string} turnkeyPublicKey key from a Turnkey iframe
+ * @returns {string} nonce to be used in OIDC
+ */
+export function getOauthNonce(turnkeyPublicKey: string): string {
+  return sha256(new TextEncoder().encode(turnkeyPublicKey)).slice(2);
+}
+
+/**
+ * Base64 URL encode function
+ *
+ * @param {ArrayBuffer | Uint8Array | string} challenge - The data to encode
+ * @returns {string} Base64 URL encoded string
+ */
+export function base64UrlEncode(
+  challenge: ArrayBuffer | Uint8Array | string
+): string {
+  let bytes: Uint8Array;
+
+  if (typeof challenge === "string") {
+    bytes = new TextEncoder().encode(challenge);
+  } else if (challenge instanceof ArrayBuffer) {
+    bytes = new Uint8Array(challenge);
+  } else {
+    bytes = challenge;
+  }
+
+  // Convert to base64 using btoa (browser-compatible)
+  let binaryString = "";
+  for (let i = 0; i < bytes.length; i++) {
+    binaryString += String.fromCharCode(bytes[i]);
+  }
+  return btoa(binaryString)
+    .replace(/\+/g, "-")
+    .replace(/\//g, "_")
+    .replace(/=/g, "");
+}
+
+/**
+ * Resolves a relative URL to an absolute URL (browser environment)
+ *
+ * @param {string} url - The URL to resolve
+ * @returns {string} The absolute URL
+ */
+export function resolveRelativeUrl(url: string): string {
+  // Funny trick.
+  const a = document.createElement("a");
```

Comment:
Since this breaks outside of browsers, ideally it would be in `signer-web`, although I don't know if that makes other parts of the code structure awkward. But we have to be sure this doesn't get called on react-native.